// atlassian-oauth2/lib/oauth2.js
var request = require('request'),
    RSVP = require('rsvp'),
    jwt = require('atlassian-jwt');

var EXPIRE_IN_SECONDS = 60,
    AUTHORIZATION_SERVER_URL = "https://auth.atlassian.io",
    JWT_CLAIM_PREFIX = "urn:atlassian:connect",
    GRANT_TYPE = "urn:ietf:params:oauth:grant-type:jwt-bearer",
    SCOPE_SEPARATOR = ' ';


/**
 * Creates a JWT claimset for authenticating the add-on to the OAuth2 service
 *
 * @param {String} hostBaseUrl - The fully qualified instance name, for example `https://instance.atlassian.net`
 * @param {String} oauthClientId - The OAuth client id which corresponds to the `hostBaseUrl` which was provided to the add-on during installation
 * @param {String} userKey - The user key (not username) of the user to retrieve an access token for
 * @param {String=} audience - The user key (not username) of the user to retrieve an access token for
 * @returns {Object} A claimset to be encoded and sent with the token request
 */
function _createAssertionPayload (hostBaseUrl, oauthClientId, userKey, audience) {
    var now = Math.floor(Date.now() / 1000);
    var exp = now + EXPIRE_IN_SECONDS;

    return {
        "iss": JWT_CLAIM_PREFIX + ":clientid:" + oauthClientId,
        "sub": JWT_CLAIM_PREFIX + ":userkey:" + userKey,
        "tnt": hostBaseUrl,
        "aud": audience || AUTHORIZATION_SERVER_URL,
        "iat": now,
        "exp": exp
    }
}

/**
 * Retrieves an OAuth 2 access token for a given user and instance by creating a JWT token
 * signed by the add-on's shared secret.
 *
 * @param {Object} opts - A hash defining the parameters of the access token request
 * @param {String} opts.hostBaseUrl - The fully qualified instance name, for example `https://instance.atlassian.net`
 * @param {String} opts.oauthClientId - The OAuth client id which corresponds to the `hostBaseUrl` which was provided to the add-on during installation
 * @param {String} opts.sharedSecret - The shared secret which corresponds to the `hostBaseUrl` which was provided to the add-on during installation
 * @param {String} opts.userKey - The user key (not username) of the user to retrieve an access token for
 * @param {String} opts.scopes - An array of scopes to request for when creating the access token
 * @param {String=} opts.authorizationServerBaseUrl - An alternative authorization server to use (intended for internal use by Atlassian only)
 * @param {String=} opts.authorizationPath - An alternative authorization path to use (intended for internal use by Atlassian only)
 * @returns {Promise.<Object, Error>} A promise that returns the access token if resolved, or an error if rejected
 */
function getAccessToken (opts) {
    opts = opts || {};
    return new RSVP.Promise(function (resolve, reject) {
        var jwtClaims = _createAssertionPayload(opts.hostBaseUrl, opts.oauthClientId, opts.userKey, opts.authorizationServerBaseUrl);
        var assertion = jwt.encode(jwtClaims, opts.sharedSecret);

        var formData = {
            grant_type: GRANT_TYPE,
            assertion: assertion
        };

        if (opts.scopes) {
            formData.scope = opts.scopes.join(SCOPE_SEPARATOR).toUpperCase()
        }

        request({
            method: 'POST',
            url: (opts.authorizationServerBaseUrl || AUTHORIZATION_SERVER_URL) + (opts.authorizationPath || '/oauth2/token'),
            form: formData,
            json: true,
            headers: {
                "accept": "application/json"
            }
        }, function(err, response, body) {
            if (err) {
                reject(err);
            } else if (response.statusCode < 200 || response.statusCode > 299) {
                reject(body);
            } else {
                resolve(body);
            }
        });
    });
}


module.exports = {
    _createAssertionPayload: _createAssertionPayload,
    getAccessToken: getAccessToken
};
